lit_stablelm/model.py

```diff
@@ -0,0 +1,220 @@
+        self.transformer = nn.ModuleDict(
+            dict(
+                wte=nn.Embedding(config.padded_vocab_size, config.n_embd),
+                h=nn.ModuleList(Block(config) for _ in range(config.n_layer)),
+                ln_f=nn.LayerNorm(config.n_embd),
+            )
+        )
+
+    def _init_weights(self, module: nn.Module) -> None:
+        if isinstance(module, nn.Linear):
+            # https://huggingface.co/stabilityai/stablelm-base-alpha-3b/blob/main/config.json#L10
+            torch.nn.init.normal_(module.weight, mean=0.0, std=0.02)
+            if module.bias is not None:
+                torch.nn.init.zeros_(module.bias)
+        elif isinstance(module, nn.Embedding):
+            torch.nn.init.normal_(module.weight, mean=0.0, std=0.02)
+        elif isinstance(module, nn.LayerNorm):
+            torch.nn.init.ones_(module.weight)
+            torch.nn.init.zeros_(module.bias)
+            # https://huggingface.co/stabilityai/stablelm-base-alpha-3b/blob/main/config.json#L12
+            module.eps = 1e-5
+
+    def forward(self, idx: torch.Tensor) -> torch.Tensor:
+        _, t = idx.size()
+        assert (
+            t <= self.config.block_size
+        ), f"Cannot forward sequence of length {t}, block size is only {self.config.block_size}"
+
+        # forward the LLaMA model itself
+        x = self.transformer.wte(idx)  # token embeddings of shape (b, t, n_embd)
+
+        for block in self.transformer.h:
+            x = block(x)
+        x = self.transformer.ln_f(x)
+
+        logits = self.lm_head(x)  # (b, t, vocab_size)
+
+        return logits
+
+    @classmethod
+    def from_name(cls, name: str) -> Self:
+        return cls(StableLMConfig.from_name(name))
+
+
+class Block(nn.Module):
+    def __init__(self, config: StableLMConfig) -> None:
+        super().__init__()
+        self.norm_1 = nn.LayerNorm(config.n_embd)
+        self.attn = CausalSelfAttention(config)
+        self.norm_2 = nn.LayerNorm(config.n_embd)
+        self.mlp = MLP(config)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        # gpt-j residual
+        x = x + self.attn(self.norm_1(x)) + self.mlp(self.norm_2(x))
+        return x
+
+
+class CausalSelfAttention(nn.Module):
+    def __init__(self, config: StableLMConfig) -> None:
+        super().__init__()
+        assert config.n_embd % config.n_head == 0
+
+        # key, query, value projections for all heads, but in a batch
+        self.attn = nn.Linear(config.n_embd, 3 * config.n_embd, bias=True)
+        # output projection
+        self.proj = nn.Linear(config.n_embd, config.n_embd, bias=True)
+
+        self.n_head = config.n_head
+        self.n_embd = config.n_embd
+        self.block_size = config.block_size
+        self.rotary_percentage = config.rotary_percentage
+        self.rope_cache: Optional[Tuple[torch.Tensor, torch.Tensor]] = None
+
+        if self.rotary_percentage != 1.0:
```

Comment:
key change: the non-flash path is used for the stablelm checkpoints because they specify a rotary percentage of 0.25